File: python/examples/ex03_sessions.py
"""Example 03 -- stateful sessions: create -> multi-step predict -> info/reset -> DELETE.

Purpose
    Sessions keep the action trajectory SERVER-side (``max_trajectory_length``
    recent steps), so every step only uploads the new screenshot -- cheaper
    and simpler than re-sending a local trajectory to /predict. This example
    runs a multi-step predict loop inside one session, shows the session info
    (``GET /v1/sessions/{id}``), demonstrates a free trajectory reset, and
    ALWAYS deletes the session in a ``finally`` block (deleting frees your
    account's concurrency slot, even when a step blew up).

Flow
    1. POST /v1/sessions pinned to the 1280x720 capture size.
    2. Loop: capture -> POST /v1/sessions/{id}/predict -> execute actions
       (same injected capture/backend pattern as ex01; coordinates scale back
       to the real screen). Each step sends a unique ``Idempotency-Key`` so a
       network retry can never double-charge or double-predict the step.
    3. GET /v1/sessions/{id} -- step_count / total_credits_used.
    4. POST /v1/sessions/{id}/reset -- clears the server trajectory (free).
    5. finally: DELETE /v1/sessions/{id} -- always, error or not (free).

Endpoints used
    POST /v1/sessions, POST /v1/sessions/{id}/predict, GET /v1/sessions/{id},
    POST /v1/sessions/{id}/reset, DELETE /v1/sessions/{id} (scope ``session``)

Estimated cost
    10 credits flat to create the session (no surcharges) + 4 credits per
    step at 1280x720 (the server-side trajectory does NOT add the +2 cr/shot
    /predict surcharge; HD would add +1 but 1280x720 is SD). Info, reset and
    delete are free. The itemized estimate is COMPUTED via :mod:`coasty.cost`
    and printed before any spend; sandbox keys print "$0 (sandbox)".

Run it
    python examples/ex03_sessions.py "Fill in the login form" --max-steps 5 --confirm
"""

from __future__ import annotations

import argparse
import sys
from collections.abc import Callable, Sequence
from dataclasses import dataclass
from typing import cast

from coasty import (
    ActionBackend,
    ActionExecutor,
    CoastyClient,
    CoastyError,
    PyAutoGuiBackend,
    cost,
    env,
)
from coasty.types import CuaVersion, PredictStatus, SessionInfo
from ex01_local_predict_loop import (
    TARGET_HEIGHT,
    TARGET_WIDTH,
    CaptureFn,
    FinalStatus,
    StepRecord,
    make_local_capture,
    spend_gate,
)


@dataclass(frozen=True)
class SessionLoopResult:
    """Outcome of one session lifecycle."""

    session_id: str
    status: FinalStatus
    steps: list[StepRecord]
    info: SessionInfo | None  # GET /sessions/{id} snapshot taken after the loop

    @property
    def total_credits(self) -> int:
        return sum(record.credits_charged for record in self.steps)


def run_session(
    client: CoastyClient,
    capture: CaptureFn,
    backend: ActionBackend,
    instruction: str,
    *,
    max_steps: int = 10,
    cua_version: CuaVersion = "v3",
    max_trajectory_length: int = 3,
    inspect_and_reset: bool = True,
    emit: Callable[[str], None] = print,
) -> SessionLoopResult:
    """The pure core: full session lifecycle with a guaranteed DELETE.

    ``capture``/``backend`` are injected (tests: fake capture + NullBackend).
    Whatever happens after create -- predict errors, executor errors -- the
    ``finally`` block deletes the session so the concurrency slot is freed.
    """
    if max_steps < 1:
        raise ValueError("max_steps must be >= 1")

    first_shot = capture()  # the session pins this screen size at create time
    created = client.create_session(
        cua_version=cua_version,
        screen_width=first_shot.sent_width,
        screen_height=first_shot.sent_height,
        max_trajectory_length=max_trajectory_length,
    ).data
    session_id = created["session_id"]
    emit(f"session {session_id} created (screen={created['screen_size']}, 10 cr flat)")

    try:
        steps: list[StepRecord] = []
        final_status: FinalStatus = "max_steps"
        shot = first_shot
        for step_number in range(1, max_steps + 1):
            result = client.session_predict(
                session_id,
                shot.screenshot_b64,
                instruction,
                # Unique per step: a 429/5xx/transport retry replays instead
                # of double-executing (and double-charging) the step.
                idempotency_key=f"{session_id}-step-{step_number}",
            )
            prediction = result.data
            status: PredictStatus = prediction["status"]
            executor = ActionExecutor(backend, scale_x=shot.scale_x, scale_y=shot.scale_y)
            executed = executor.execute_all(prediction["actions"])
            steps.append(
                StepRecord(
                    step=prediction["step"],
                    status=status,
                    request_id=prediction["request_id"],
                    executed=executed,
                    credits_charged=prediction["usage"]["credits_charged"],
                )
            )
            emit(
                f"step {prediction['step']}: status={status} actions={executed} "
                f"({steps[-1].credits_charged} cr, request_id={steps[-1].request_id})"
            )
            if status != "continue":
                final_status = status
                break
            shot = capture()
        else:
            emit(f"stopping: --max-steps {max_steps} reached before done/fail")

        info: SessionInfo | None = None
        if inspect_and_reset:
            info = client.get_session(session_id).data  # free
            emit(
                f"session info: step_count={info['step_count']} "
                f"total_credits_used={info['total_credits_used']} expires_at={info['expires_at']}"
            )
            client.reset_session(session_id)  # free: wipes the server trajectory
            emit("session trajectory reset (free) -- next predict starts fresh")
        return SessionLoopResult(session_id=session_id, status=final_status, steps=steps, info=info)
    finally:
        # ALWAYS delete -- even when a step raised -- to free the concurrency
        # slot. A delete failure is reported but never masks the original error.
        try:
            client.delete_session(session_id)  # free
            emit(f"session {session_id} deleted (concurrency slot freed)")
        except CoastyError as exc:
            emit(f"warning: failed to delete session {session_id}: {exc}")


def build_estimate(max_steps: int, cua_version: CuaVersion) -> cost.CostEstimate:
    """Itemized worst case: one create + ``max_steps`` session predicts."""
    per_step = cost.estimate_session_predict(
        cua_version=cua_version, screen_width=TARGET_WIDTH, screen_height=TARGET_HEIGHT
    )
    scaled_steps = cost.CostEstimate(
        items=tuple(
            cost.CostItem(f"{item.label} x{max_steps} step(s)", item.credits * max_steps)
            for item in per_step.items
        )
    )
    return cost.combine(cost.estimate_session_create(), scaled_steps)


def main(argv: Sequence[str] | None = None) -> int:
    parser = argparse.ArgumentParser(
        description="Stateful session lifecycle: create -> predict loop -> info/reset -> delete."
    )
    parser.add_argument("instruction", help="natural-language task for the session")
    parser.add_argument("--max-steps", type=int, default=10, help="hard cap on session steps")
    parser.add_argument("--cua-version", choices=("v1", "v3", "v4"), default="v3")
    parser.add_argument(
        "--max-trajectory-length", type=int, default=3, help="server-kept steps (1-20)"
    )
    parser.add_argument(
        "--confirm", action="store_true", help="allow spending on a live (non-sandbox) key"
    )
    args = parser.parse_args(list(argv) if argv is not None else None)
    cua_version = cast(CuaVersion, args.cua_version)

    try:
        api_key = env.require_api_key()
    except env.MissingAPIKeyError as exc:
        print(f"error: {exc}", file=sys.stderr)
        return 2

    estimate = build_estimate(args.max_steps, cua_version)
    if not spend_gate(
        estimate, api_key=api_key, confirm=args.confirm, title="ex03 session lifecycle"
    ):
        return 2

    try:
        capture = make_local_capture()
        backend = PyAutoGuiBackend(failsafe=True)  # corner-of-screen abort stays on
    except RuntimeError as exc:
        print(f"error: {exc}", file=sys.stderr)
        return 2

    try:
        with CoastyClient(api_key) as client:
            outcome = run_session(
                client,
                capture,
                backend,
                args.instruction,
                max_steps=args.max_steps,
                cua_version=cua_version,
                max_trajectory_length=args.max_trajectory_length,
            )
    except CoastyError as exc:
        print(f"error: {exc}", file=sys.stderr)  # str() includes the request_id
        if exc.request_id:
            print(f"request_id: {exc.request_id}", file=sys.stderr)
        return 1

    print(
        f"finished: {outcome.status} after {len(outcome.steps)} step(s); "
        f"{outcome.total_credits} credit(s) charged on session {outcome.session_id}"
    )
    return 0 if outcome.status == "done" else 1


if __name__ == "__main__":
    sys.exit(main())

File: typescript/src/examples/ex03-sessions.ts
/**
 * Example 03 — Stateful sessions: create -> multi-step predict -> delete.
 *
 * Purpose: sessions keep the trajectory (recent screenshots + actions)
 * SERVER-side, so each step is a smaller request AND cheaper than stateless
 * /predict with a client-side trajectory (4 cr vs 5 cr + 2 cr per attached
 * trajectory screenshot). This example walks the full lifecycle:
 *
 *   1. POST /v1/sessions                     (create — 10 cr one-time)
 *   2. POST /v1/sessions/{id}/predict  x N   (4 cr each at <=1280x720)
 *   3. GET  /v1/sessions/{id}                (info: step_count, credits — free)
 *   4. POST /v1/sessions/{id}/reset          (clear trajectory, keep slot — free)
 *   5. DELETE /v1/sessions/{id}              (ALWAYS, in `finally` — free)
 *
 * The delete is in a `finally` block on purpose: sessions hold a concurrency
 * slot until they expire, so leaking them on errors throttles future work.
 *
 * The screen is driven through the same injected screenshot-provider +
 * ExecutorBackend pattern as example 01 (optional Playwright in `main()`).
 *
 * Estimated cost (computed at runtime via src/coasty/cost.ts and printed by
 * the spend gate): estimateSessionCreateCredits() + maxSteps x
 * estimateSessionPredictCredits({1280, 720}) = 10 + 6 x 4 = 34 cr = $0.34
 * ceiling with the defaults. Sandbox keys never bill ("$0 (sandbox)").
 *
 * Run it:
 *   npx tsx src/examples/ex03-sessions.ts \
 *     --url https://example.com --instruction "Click the More information link" \
 *     --max-steps 6 --confirm
 *
 * Env config: COASTY_API_KEY (required), COASTY_BASE_URL, EX03_URL,
 * EX03_INSTRUCTION, EX03_MAX_STEPS, COASTY_CONFIRM_SPEND=1.
 */
import { randomUUID } from 'node:crypto';
import { CoastyClient } from '../coasty/client.js';
import {
  estimateSessionCreateCredits,
  estimateSessionPredictCredits,
  formatUsd,
} from '../coasty/cost.js';
import { getApiKey } from '../coasty/env.js';
import { executeActions, type ExecutorBackend } from '../coasty/executor.js';
import { type CuaVersion } from '../coasty/types.js';
import {
  SCREEN_HEIGHT,
  SCREEN_WIDTH,
  createPageBackend,
  createPageScreenshotProvider,
  describeError,
  ensureSpendApproved,
  isCliEntry,
  openPlaywrightPage,
  stdoutPrint,
  type PredictLoopStatus,
  type PrintFn,
  type ScreenshotProvider,
} from './ex01-local-predict-loop.js';

// ---------------------------------------------------------------------------
// Core (pure: everything injected, fully testable offline)
// ---------------------------------------------------------------------------

export interface SessionLoopOptions {
  client: CoastyClient;
  instruction: string;
  screenshot: ScreenshotProvider;
  backend: ExecutorBackend;
  /** Hard cap on session predict calls (default 6). */
  maxSteps?: number;
  screenWidth?: number;
  screenHeight?: number;
  cuaVersion?: CuaVersion;
  /** Server-side trajectory window, 1-20 (default 3). */
  maxTrajectoryLength?: number;
  /** Also demonstrate GET info + POST reset after the loop (default true). */
  showLifecycle?: boolean;
  print?: PrintFn;
}

export interface SessionLoopResult {
  sessionId: string;
  status: PredictLoopStatus;
  /** Number of /sessions/{id}/predict calls made. */
  stepsUsed: number;
  /** Client-side sum of usage.credits_charged (excludes the 10 cr create fee). */
  creditsCharged: number;
  /** step_count reported by GET /sessions/{id} (null when showLifecycle=false). */
  stepCountReported: number | null;
  /** total_credits_used reported by GET /sessions/{id}. */
  totalCreditsReported: number | null;
  failReason: string | null;
}

/**
 * Full session lifecycle. The session is ALWAYS deleted — on success AND
 * when a predict throws mid-loop (finally semantics, structured as an
 * explicit catch + rethrow so a cleanup failure during error unwind is
 * printed, never thrown, and the original error is what callers see).
 */
export async function runSessionLoop(options: SessionLoopOptions): Promise<SessionLoopResult> {
  const maxSteps = options.maxSteps ?? 6;
  const print = options.print ?? stdoutPrint;

  const { data: created } = await options.client.sessions.create({
    cua_version: options.cuaVersion ?? 'v3',
    screen_width: options.screenWidth ?? SCREEN_WIDTH,
    screen_height: options.screenHeight ?? SCREEN_HEIGHT,
    max_trajectory_length: options.maxTrajectoryLength ?? 3,
  });
  const sessionId = created.session_id;
  print(
    `created session ${sessionId} (cua_version=${created.cua_version}, ` +
      `screen=${created.screen_size}, expires_at=${created.expires_at})`,
  );

  // ALWAYS delete: sessions hold a concurrency slot until they expire.
  // `suppressError` is true on the error-unwind path so a cleanup failure
  // can never mask the original error (it is printed instead).
  const deleteSession = async (suppressError: boolean): Promise<void> => {
    try {
      await options.client.sessions.delete(sessionId);
      print(`deleted session ${sessionId} (concurrency slot freed).`);
    } catch (deleteError) {
      if (!suppressError) throw deleteError;
      print(`warning: failed to delete session ${sessionId}: ${describeError(deleteError)}`);
    }
  };

  let result: SessionLoopResult;
  try {
    let status: PredictLoopStatus = 'max_steps';
    let stepsUsed = 0;
    let creditsCharged = 0;
    let failReason: string | null = null;

    for (let step = 1; step <= maxSteps; step += 1) {
      const screenshot = await options.screenshot();
      const { data } = await options.client.sessions.predict(
        sessionId,
        { screenshot, instruction: options.instruction },
        // Session predicts advance server-side trajectory state, so they are
        // NOT inherently retry-safe; an Idempotency-Key opts into retries.
        { idempotencyKey: `ex03-step${String(step)}-${randomUUID()}` },
      );
      stepsUsed = step;
      creditsCharged += data.usage.credits_charged;
      print(
        `[step ${String(data.step)}] status=${data.status} actions=${String(data.actions.length)} ` +
          `credits=${String(data.usage.credits_charged)}` +
          (data.reasoning !== null ? ` reasoning=${data.reasoning}` : ''),
      );

      const results = await executeActions(data.actions, options.backend, { logger: print });
      const terminal = results.find((result) => result.terminal !== null) ?? null;
      if (data.status === 'fail' || terminal?.terminal === 'fail') {
        status = 'fail';
        failReason = terminal?.detail ?? data.reasoning;
        break;
      }
      if (data.status === 'done' || terminal?.terminal === 'done') {
        status = 'done';
        break;
      }
    }

    let stepCountReported: number | null = null;
    let totalCreditsReported: number | null = null;
    if (options.showLifecycle ?? true) {
      // GET info is free: the server's own view of steps + credits spent.
      const { data: info } = await options.client.sessions.get(sessionId);
      stepCountReported = info.step_count;
      totalCreditsReported = info.total_credits_used;
      print(
        `session info: step_count=${String(info.step_count)}, ` +
          `total_credits_used=${String(info.total_credits_used)} ` +
          `(${formatUsd(info.total_credits_used)})`,
      );
      // Reset (free) clears the trajectory but keeps the session: start the
      // next task on the same session without paying the 10 cr create fee.
      await options.client.sessions.reset(sessionId);
      print('session reset: trajectory cleared, session (and its slot) kept.');
    }

    result = {
      sessionId,
      status,
      stepsUsed,
      creditsCharged,
      stepCountReported,
      totalCreditsReported,
      failReason,
    };
  } catch (error) {
    await deleteSession(true); // cleanup runs even though the loop threw
    throw error;
  }
  await deleteSession(false); // success path: a delete failure is a real error
  return result;
}

// ---------------------------------------------------------------------------
// CLI
// ---------------------------------------------------------------------------

export interface Ex03Config {
  url: string;
  instruction: string;
  maxSteps: number;
  headless: boolean;
  confirm: boolean;
}

function intOption(name: string, raw: string | undefined, fallback: number): number {
  if (raw === undefined || raw.trim() === '') return fallback;
  const value = Number(raw);
  if (!Number.isInteger(value) || value < 1) {
    throw new Error(`${name} must be a positive integer, got: ${raw}`);
  }
  return value;
}

export function parseArgs(
  argv: readonly string[],
  env: NodeJS.ProcessEnv = process.env,
): Ex03Config {
  const config: Ex03Config = {
    url: env.EX03_URL ?? 'https://example.com',
    instruction: env.EX03_INSTRUCTION ?? 'Click the "More information..." link.',
    maxSteps: intOption('EX03_MAX_STEPS', env.EX03_MAX_STEPS, 6),
    headless: true,
    confirm: false,
  };
  for (let i = 0; i < argv.length; i += 1) {
    const arg = argv[i];
    const next = (): string => {
      i += 1;
      const value = argv[i];
      if (value === undefined) throw new Error(`missing value for ${String(arg)}`);
      return value;
    };
    switch (arg) {
      case '--url':
        config.url = next();
        break;
      case '--instruction':
      case '-i':
        config.instruction = next();
        break;
      case '--max-steps':
        config.maxSteps = intOption('--max-steps', next(), config.maxSteps);
        break;
      case '--headed':
        config.headless = false;
        break;
      case '--confirm':
        config.confirm = true;
        break;
      default:
        throw new Error(
          `unknown argument: ${String(arg)} (expected --url, --instruction, --max-steps, --headed, --confirm)`,
        );
    }
  }
  return config;
}

export async function main(argv: readonly string[] = process.argv.slice(2)): Promise<number> {
  const config = parseArgs(argv);
  const apiKey = getApiKey();
  const perStep = estimateSessionPredictCredits({
    screenWidth: SCREEN_WIDTH,
    screenHeight: SCREEN_HEIGHT,
  });
  const approved = ensureSpendApproved({
    apiKey,
    confirmFlag: config.confirm,
    items: [
      { label: '1 x POST /sessions (create)', credits: estimateSessionCreateCredits() },
      {
        label: `${String(config.maxSteps)} x POST /sessions/{id}/predict @ ${String(SCREEN_WIDTH)}x${String(SCREEN_HEIGHT)}`,
        credits: config.maxSteps * perStep,
      },
      { label: 'get + reset + delete', credits: 0 },
    ],
  });
  if (!approved) return 1;

  const client = new CoastyClient({ apiKey });
  const { page, close } = await openPlaywrightPage({ url: config.url, headless: config.headless });
  try {
    const result = await runSessionLoop({
      client,
      instruction: config.instruction,
      screenshot: createPageScreenshotProvider(page),
      backend: createPageBackend(page),
      maxSteps: config.maxSteps,
    });
    stdoutPrint(
      `result: ${result.status} after ${String(result.stepsUsed)} step(s), ` +
        `${String(result.creditsCharged)} cr in predicts (${formatUsd(result.creditsCharged)})`,
    );
    return result.status === 'fail' ? 1 : 0;
  } finally {
    await close();
  }
}

if (isCliEntry(import.meta.url)) {
  void main()
    .then((code) => {
      process.exitCode = code;
    })
    .catch((error: unknown) => {
      console.error(describeError(error));
      process.exitCode = 1;
    });
}
